runtime/compiler/control/CompilationThread.cpp

```diff
@@ -6028,6 +6033,48 @@ void *TR::CompilationInfo::compileOnSeparateThread(J9VMThread * vmThread, TR::Il
       if (compErrCode)
          *compErrCode = (TR_CompilationErrorCode)entry->_compErrCode;
 
+      // If a remote sync compilation has been changed to a local sync compilation,
```

Comment:
I wonder if we need to put all this code somewhat below when we acquire comp monitor. There is this part of the code
```
      // Release the monitor now
      //
      debugPrint(vmThread, "\tapplication thread releasing AppMonitor\n");
      debugPrint(vmThread, "-AM-", entry);

      entry->releaseSlotMonitor(vmThread);

      // Grab the compilation mutex to be able to decrement numThreadsWaiting
      // and access entryShouldBeDeallocated
      // When numThreadsWaiting reaches 0, the entry can be reused or deallocated
      //
      acquireCompMonitor(vmThread);
      entry->_numThreadsWaiting--;
```
which suggests that `_numThreadsWaiting` is protected by the compilation mutex and when you test it we don't have it locked. This change also prevents us from acquiring the comp monitor then releasing it and then acquiring again.

Reply:
I can move `acquireCompMonitor` and `releaseCompMonitor` outside the `if` condition check so that checking `_numThreadsWaiting` is guarded with locks. Or pull the code before `entry->_numThreadsWaiting--`. Let me test it and see how it goes.